server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -362,26 +362,38 @@ protected Pool assembleDevPool(Consumer consumer, Owner owner, String sku) {
         return pool;
     }
 
-    private DeveloperProducts getDeveloperPoolProducts(Consumer consumer, Owner owner, String sku) {
-        DeveloperProducts devProducts = getDevProductMap(consumer, owner, sku);
-        verifyDevProducts(consumer, sku, devProducts);
+    private DeveloperProducts getDeveloperPoolProducts(Owner owner, String sku) {
+        DeveloperProducts devProducts = getDevProductMap(owner, sku);
+        verifyDevProducts(sku, devProducts);
         return devProducts;
     }
 
     /**
-     * Looks up all Products matching the specified SKU and the consumer's
-     * installed products.
+     * Looks up all Products and their provided products
+     * matching the specified SKU.
      *
-     * @param consumer the consumer to pull the installed product id list from.
      * @param sku the product id of the SKU.
      * @return a {@link DeveloperProducts} object that contains the Product objects
      *         from the adapter.
      */
-    private DeveloperProducts getDevProductMap(Consumer consumer, Owner owner, String sku) {
-        List<String> devProductIds = new ArrayList<>();
-        devProductIds.add(sku);
-        for (ConsumerInstalledProduct ip : consumer.getInstalledProducts()) {
-            devProductIds.add(ip.getProductId());
+    private DeveloperProducts getDevProductMap(Owner owner, String sku) {
+
+        Collection<? extends ProductInfo> productsByIds = this.productAdapter
+            .getProductsByIds(owner.getKey(), Arrays.asList(sku));
+
+        Collection<String> devProductIds = new ArrayList<>();
+        if (productsByIds.iterator().hasNext()) {
+            ProductInfo devProduct = productsByIds.iterator().next();
+            devProductIds.add(devProduct.getId());
+
+            if (!devProduct.getProvidedProducts().isEmpty()) {
```

Comment:
This block is a bit dangerous, since we have no innate control over whether or not devProduct actually returns a non-null value or not. Plus, we can do some optimization here to avoid instantiating a new collection, since we already have one to populate:

```
            Collection<? extends ProductInfo> provided = devProduct.getProvidedProducts();

            if (provided != null) {
                provided.forEach(product -> {
                    if (product != null) {
                        devProductIds.add(product.getId());
                    }
                });
            }
```

Reply:
Agreed, null objects needed to be filtered out. Updated as suggested.